cbmc/proofs/IotMqtt_SubscribeAsync/IotMqtt_SubscribeAsync_harness.c

```diff
@@ -0,0 +1,107 @@
+#include "iot_config.h"
+#include "private/iot_mqtt_internal.h"
+
+#include <stdlib.h>
+
+#include "mqtt_state.h"
+
+/**
+ * We abstract all functions related to concurrency and assume they are correct.
+ * Thus, we add these stub to increase coverage.
+ */
+IotTaskPoolError_t IotTaskPool_CreateJob( IotTaskPoolRoutine_t userCallback,
+                                          void * pUserContext,
+                                          IotTaskPoolJobStorage_t * const pJobStorage,
+                                          IotTaskPoolJob_t * const pJob )
+{
+  if (userCallback == NULL || pJobStorage == NULL || pJob == NULL)
+  {
+    return IOT_TASKPOOL_BAD_PARAMETER;
+  }
+  /* _IotMqtt_ScheduleOperation asserts this */
+  return IOT_TASKPOOL_SUCCESS;
+}
+
+IotTaskPoolError_t IotTaskPool_ScheduleDeferred( IotTaskPool_t taskPool,
+                                                 IotTaskPoolJob_t job,
+                                                 uint32_t timeMs )
+{
+  IotTaskPoolError_t error;
+
+  /* _IotMqtt_ScheduleOperation asserts this */
+  __CPROVER_assume(error != IOT_TASKPOOL_BAD_PARAMETER &&
+                   error != IOT_TASKPOOL_ILLEGAL_OPERATION);
+  return error;
+}
+
+/**
+ * _IotMqtt_NextPacketIdentifier calls Atomic_Add_u32, which receives
+ * a volatile variable as input. Thus, CBMC will always consider that
+ * Atomic_Add_u32 will operate over nondetermistic values and raises
+ * a unsigned integer overflow failure. However, developers have reported
+ * that the use of this overflow is part of the function implementation.
+ * In order to mirror _IotMqtt_NextPacketIdentifier behaviour and avoid
+ * spurious alarms, we stub out this function to always
+ * return a nondetermistic odd value.
+ */
+uint16_t _IotMqtt_NextPacketIdentifier( void ) {
+  uint16_t id;
+
+  /* Packet identifiers will follow the sequence 1,3,5...65535,1,3,5... */
+  __CPROVER_assume(id & 0x0001 == 1);
+
+  return id;
+}
+
```

Comment:
We are using these abstractions four places now, I believe, we should discuss at what point it makes sense to move them into a common library.  Doing so, however, would require a change to our build system.

Reply:
I totally agree. In the past, we had the idea that a proof harness should contain all the code (i.e., assumptions) necessary for the proof, which included the stubs implementation. But, if we need to update one stub, than we must go through each proof fixing them. I'll open a GitHub issue, so we handle this on a separate/dedicated PR.

Reply:
Issue https://github.com/aws/aws-iot-device-sdk-embedded-C/issues/829 will track this request.